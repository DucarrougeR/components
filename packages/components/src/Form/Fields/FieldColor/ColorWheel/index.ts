export * from './ColorWheel'
export { HueSaturation } from './color_wheel_utils'
