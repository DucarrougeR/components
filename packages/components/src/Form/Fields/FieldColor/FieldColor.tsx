import React, { useState } from 'react'
import { Popover, PopoverContent } from '../../../Popover'
import { InputText, InputTextProps } from '../../Inputs/InputText'
import { Field, FieldProps } from '../Field'
import { FormControl } from '../../FormControl'
import { useFormContext } from '../../Form'
import {
  HueSaturation,
  polarbrightness2hsv,
  SimpleHSV,
  white,
} from './ColorWheel/color_wheel_utils'
import { ColorWheel } from './ColorWheel'
import { LuminositySlider } from './LuminositySlider'
import { Swatch } from './Swatch'
import {
  hsv2hex,
  simpleHSVtoFormattedColorString,
  str2simpleHsv,
} from './utils/color_format_utils'
import { isValidColor } from './utils/color_utils'

const colorWheelSize = 164

export interface FieldColorProps
  extends FieldProps,
    Omit<InputTextProps, 'height' | 'width' | 'size'> {
  /**
   * If true, hides input and only show color swatch.
   */
  hideInput?: boolean

  value?: string
}

const createEventWithHSVValue = (
  color: SimpleHSV
): React.ChangeEvent<HTMLInputElement> => {
  return {
    currentTarget: {
      value: simpleHSVtoFormattedColorString(color),
    },
    target: {
      value: simpleHSVtoFormattedColorString(color),
    },
  } as React.ChangeEvent<HTMLInputElement>
}

export const FieldColor: React.FC<FieldColorProps> = ({
  alignValidationMessage,
  hideInput,
  ...props
}) => {
  const validationMessage = useFormContext(props)
  const initialValue =
    props.value && isValidColor(props.value)
      ? str2simpleHsv(props.value)
      : polarbrightness2hsv(white())

  const [color, setColor] = useState<SimpleHSV>(initialValue)
  const [inputTextValue, setInputTextValue] = useState(props.value || '')

  const callOnChange = () => {
    if (!props.onChange || !color) return
    props.onChange(createEventWithHSVValue(color))
  }

  const setColorState = (newColor: SimpleHSV) => {
    setColor(newColor)
    newColor && setInputTextValue(simpleHSVtoFormattedColorString(newColor))
    callOnChange()
  }

  const handleColorChange = (hs: HueSaturation) =>
    setColorState({ ...hs, v: color.v })

  const handleSliderChange = (event: React.FormEvent<HTMLInputElement>) =>
    setColorState({
      ...color,
      v: Number(event.currentTarget.value) / 100,
    })

  const handleInputTextChange = (event: React.FormEvent<HTMLInputElement>) => {
    const value = event.currentTarget.value
    setInputTextValue(value)

    if (!value || !isValidColor(value)) return

    setColor(str2simpleHsv(value))
    callOnChange()
  }

  const content = (
    <PopoverContent display="flex" flexDirection="column">
      <ColorWheel
        size={colorWheelSize}
        hue={color.h}
        saturation={color.s}
        value={color.v}
        onColorChange={handleColorChange}
      />
      <LuminositySlider
        min={0}
        max={100}
        step={1}
        value={color.v * 100}
        width={colorWheelSize}
        onChange={handleSliderChange}
      />
    </PopoverContent>
  )

  return (
    <Field
      {...props}
      alignValidationMessage={alignValidationMessage || 'bottom'}
    >
      <FormControl alignLabel="left">
        <Popover content={content}>
          {(onClick, ref) => (
            <Swatch
              mt="auto"
              ref={ref}
              color={hsv2hex(color)}
              borderRadius={hideInput ? 'medium' : 'none'}
              borderTopLeftRadius="medium"
              borderBottomLeftRadius="medium"
              borderRight={!hideInput ? 'none' : undefined}
              width="33px"
              height="28px"
              onClick={onClick}
            />
          )}
        </Popover>
        {!hideInput && (
          <InputText
            borderRadius="none"
            borderTopRightRadius="medium"
            borderBottomRightRadius="medium"
            validationType={validationMessage && validationMessage.type}
            onChange={handleInputTextChange}
            value={inputTextValue}
          />
        )}
      </FormControl>
    </Field>
  )
}
