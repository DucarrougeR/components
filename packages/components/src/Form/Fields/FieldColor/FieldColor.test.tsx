import 'jest-styled-components'
import React from 'react'
import {
  assertSnapshot,
  assertSnapshotShallow,
} from '@looker/components-test-utils'
import { FieldColor } from './FieldColor'

test('Default FieldColor', () => {
  assertSnapshot(<FieldColor />)
})

test('FieldColor with hidden input', () => {
  assertSnapshot(<FieldColor hideInput={true} />)
})

test('FieldColor with a named color and size values', () => {
  assertSnapshot(<FieldColor value="blue" cwSize={300} />)
})

test('FieldColor with a label', () => {
  assertSnapshot(
    <FieldColor value="#e1ff83" label="Pick a color" alignLabel="left" />
  )
})

test('FieldColor with a validation message', () => {
  assertSnapshot(
    <FieldColor
      value="#4c6670"
      label="Pick a color"
      alignLabel="left"
      validationMessage={{ message: 'Error!', type: 'error' }}
    />
  )
})

test('FieldColor renders a color picker in a Popover', () => {
  assertSnapshotShallow(<FieldColor open />)
})
