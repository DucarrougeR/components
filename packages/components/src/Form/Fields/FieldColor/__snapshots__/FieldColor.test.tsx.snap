// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default FieldColor 1`] = `
.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: 1px solid;
  border-color: #C1C6CC;
  border-radius: 0.25rem 0 0 0.25rem;
  border-right: 0;
  background-color: #ffffff;
  width: 33px;
  height: 28px;
}

.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-top: auto;
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  color: #262D33;
  margin-right: 0.5rem;
  font-weight: 600;
  font-size: 0.875rem;
}

<div
  className="c0 "
>
  <label
    className="c1"
    color="palette.charcoal800"
    fontSize="small"
    fontWeight="semiBold"
  />
  <div
    className="c0"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <div
            className="c5"
            height="28px"
            onClick={[Function]}
            width="33px"
          />
        </div>
        <div
          className="c2"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`FieldColor renders a color picker in a Popover 1`] = `ShallowWrapper {}`;

exports[`FieldColor with a label 1`] = `
.c8 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: 1px solid;
  border-color: #C1C6CC;
  border-radius: 0.25rem 0 0 0.25rem;
  border-right: 0;
  background-color: #e1ff83;
  width: 33px;
  height: 28px;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-top: auto;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  color: #262D33;
  margin-right: 0.5rem;
  font-weight: 600;
  font-size: 0.875rem;
}

.c1 .c2 {
  text-align: right;
  width: 20%;
}

<div
  className="c0 c1"
  value="#e1ff83"
>
  <label
    className="c2 c3"
    color="palette.charcoal800"
    fontSize="small"
    fontWeight="semiBold"
  >
    Pick a color
  </label>
  <div
    className="c4"
  >
    <div
      className="c5"
    >
      <div
        className="c6"
      >
        <div
          className="c7"
        >
          <div
            className="c8"
            height="28px"
            onClick={[Function]}
            width="33px"
          />
        </div>
        <div
          className="c5"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`FieldColor with a named color and size values 1`] = `
.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: 1px solid;
  border-color: #C1C6CC;
  border-radius: 0.25rem 0 0 0.25rem;
  border-right: 0;
  background-color: #0000ff;
  width: 33px;
  height: 28px;
}

.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-top: auto;
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  color: #262D33;
  margin-right: 0.5rem;
  font-weight: 600;
  font-size: 0.875rem;
}

<div
  className="c0 "
  value="blue"
>
  <label
    className="c1"
    color="palette.charcoal800"
    fontSize="small"
    fontWeight="semiBold"
  />
  <div
    className="c0"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <div
            className="c5"
            height="28px"
            onClick={[Function]}
            width="33px"
          />
        </div>
        <div
          className="c2"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`FieldColor with a validation message 1`] = `
.c8 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: 1px solid;
  border-color: #C1C6CC;
  border-radius: 0.25rem 0 0 0.25rem;
  border-right: 0;
  background-color: #4c6670;
  width: 33px;
  height: 28px;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-top: auto;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  color: #262D33;
  margin-right: 0.5rem;
  font-weight: 600;
  font-size: 0.875rem;
}

.c9 {
  color: #990F14;
  margin-right: 0.5rem;
  margin-top: 0.5rem;
  font-size: 0.75rem;
}

.c1 .c2 {
  text-align: right;
  width: 20%;
}

<div
  className="c0 c1"
  value="#4c6670"
>
  <label
    className="c2 c3"
    color="palette.charcoal800"
    fontSize="small"
    fontWeight="semiBold"
  >
    Pick a color
  </label>
  <div
    className="c4"
  >
    <div
      className="c5"
    >
      <div
        className="c6"
      >
        <div
          className="c7"
        >
          <div
            className="c8"
            height="28px"
            onClick={[Function]}
            width="33px"
          />
        </div>
        <div
          className="c5"
        />
      </div>
    </div>
    <div
      className="c9"
      fontSize="xsmall"
      type="error"
    >
      Error!
    </div>
  </div>
</div>
`;

exports[`FieldColor with hidden input 1`] = `
.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border: 1px solid;
  border-color: #C1C6CC;
  border-radius: 0.25rem 0 0 0.25rem;
  border-right: 0;
  background-color: #ffffff;
  width: 33px;
  height: 28px;
}

.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-top: auto;
}

.c6 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border-radius: 0 0.25rem 0.25rem 0;
  border: solid 1px;
  border-color: #C1C6CC;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
}

.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  margin-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  color: #262D33;
  margin-right: 0.5rem;
  font-weight: 600;
  font-size: 0.875rem;
}

<div
  className="c0 "
>
  <label
    className="c1"
    color="palette.charcoal800"
    fontSize="small"
    fontWeight="semiBold"
  />
  <div
    className="c0"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <div
            className="c5"
            height="28px"
            onClick={[Function]}
            width="33px"
          />
        </div>
        <div
          className="c2"
        >
          <input
            className="c6"
            fontSize="small"
            height="28px"
            onChange={[Function]}
            type="text"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
